eth/downloader/downloader.go

```diff
@@ -1275,15 +1286,11 @@ func (d *Downloader) fetchParts(deliveryCh chan dataPack, deliver func(dataPack)
 							// Timeouts can occur if e.g. compaction hits at the wrong time, and can be ignored
 							peer.log.Warn("Downloader wants to drop peer, but peerdrop-function is not set", "peer", pid)
 						} else {
+							// In dropPeer function, a callback will be called which aborts
+							// the sync immediately if the unregisted peer is master peer.
+							// If the peer is master one, return concrete error here.
 							d.dropPeer(pid)
-
-							// If this peer was the master peer, abort sync immediately
-							d.cancelLock.RLock()
-							master := pid == d.cancelPeer
-							d.cancelLock.RUnlock()
-
-							if master {
-								d.cancel()
+							if d.isMaster(pid) {
```

Comment:
With this change, you are introducing a gap between checking `isMaster` and actually doing the `cancel`. I assume that the `cancel` now happens later, during `UnregisterPeer` -- but is it possible that when that happens, the `pid` is no longer the master?

Reply:
`d.cancelPeer` is only assigned before syncing start at `d.synchronise` function. 

Although we drop the peer but `d.synchronising` is still 1, so we don't need to worry about the mismatch

Reply:
>  I assume that the `cancel` now happens later

Actually `cancel` now happen earlier in `dropPeer` function